pkg/apis/query/v1alpha1/cache_types.go

```diff
@@ -0,0 +1,151 @@
+type ClusterCacheSpec struct {
+	// ClusterSelectors represents the filter to select clusters.
+	// +required
+	ClusterSelectors []ClusterSelector `json:"clusterSelectors"`
+
+	// ResourceSelectors used to select resources.
+	// +required
+	ResourceSelectors []ResourceSelector `json:"resourceSelectors"`
+
+	// StatusUpdatePeriodSeconds is the period to update the status of the resource.
+	// default is 10s.
+	// +optional
+	StatusUpdatePeriodSeconds uint32 `json:"statusUpdatePeriodSeconds,omitempty"`
+}
+
+// ClusterSelector represents the filter to select clusters.
+type ClusterSelector struct {
+	// LabelSelector is a filter to select member clusters by labels.
+	// If non-nil and non-empty, only the clusters match this filter will be selected.
+	// +optional
+	LabelSelector *metav1.LabelSelector `json:"labelSelector,omitempty"`
+
+	// FieldSelector is a filter to select member clusters by fields.
+	// If non-nil and non-empty, only the clusters match this filter will be selected.
+	// +optional
+	FieldSelector *FieldSelector `json:"fieldSelector,omitempty"`
+
+	// ClusterNames is the list of clusters to be selected.
+	// +optional
+	ClusterNames []string `json:"clusterNames,omitempty"`
+
+	// ExcludedClusters is the list of clusters to be ignored.
+	// +optional
+	ExcludeClusters []string `json:"exclude,omitempty"`
+}
+
+// FieldSelector is a field filter.
+type FieldSelector struct {
+	// A list of field selector requirements.
+	MatchExpressions []corev1.NodeSelectorRequirement `json:"matchExpressions,omitempty"`
+}
+
+// ResourceSelector the resources will be selected.
+type ResourceSelector struct {
+	// APIVersion represents the API version of the target resources.
+	// +required
+	APIVersion string `json:"apiVersion"`
+
+	// Kind represents the Kind of the target resources.
+	// +required
+	Kind string `json:"kind"`
+
+	// Namespace of the target resource.
+	// Default is empty, which means all namespaces.
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
+}
+
+// ClusterCacheStatus defines the observed state of ClusterCache
+type ClusterCacheStatus struct {
+	// +optional
+	ResourceStatuses []ResourceStatusRef `json:"resources,omitempty"`
+	// +optional
+	StartTime *metav1.Time `json:"startTime,omitempty"`
+}
+
+// ResourceStatusRef is the status of the resource.
+type ResourceStatusRef struct {
+	// +required
+	Cluster string `json:"cluster"`
+	// +optional
+	APIVersion string `json:"apiVersion,omitempty"`
+	// +optional
+	Kind string `json:"kind"`
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
+	// +required
+	State CachePhase `json:"state"`
+	// +required
+	TotalNum int32 `json:"totalNum"`
+	// +required
+	UpdateTime *metav1.Time `json:"updateTime"`
+}
+
+// CachePhase is the current state of the cache
+// +enum
+type CachePhase string
+
+// These are the valid statuses of cache.
+const (
+	CacheRunning CachePhase = "Running"
+	CacheFailed  CachePhase = "Failed"
+	CacheUnknown CachePhase = "Unknown"
+)
```

Comment:
I suppose we can remove the `CachePhase` filed for now, because we are not sure if we really need it yet.
And on the other hand, we do need another filed to explain the failed reason for `Failed` and `Unknown`.

In addition, according to [Kubernetes API convention](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status), the pattern of using `phase` is deprecated: 
> Some resources in the v1 API contain fields called phase, and associated message, reason, and other status fields. The pattern of using phase is deprecated. Newer API types should use conditions instead. Phase was essentially a state-machine enumeration field, that contradicted [system-design principles](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/architecture/principles.md#control-logic) and hampered evolution, since [adding new enum values breaks backward compatibility](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md).

We can add more `Conditions` once there is a need.